resource/src/lib.rs

```diff
@@ -22,201 +23,134 @@ pub const CKB_CONFIG_FILE_NAME: &str = "ckb.toml";
     }
 
-    /// Resolves a resource using a path as the path is refered in the resource `relative_to`.
-    ///
-    /// This function is similar to [`ResourceLocator::resolve`]. The difference is how to resolve a relative path.
-    ///
-    /// [`ResourceLocator::resolve`]: struct.ResourceLocator.html#method.open
-    ///
-    /// The relative path is relative to the directory containing the resource `relative_to`.
-    ///
-    /// For security reason, when `relative_to` is `Resource::Bundled`, the return value is either
-    /// `Some(Resource::Bundled)` or `None`. A bundled file is forbidden to reference a file in the
-    /// file system.
-    pub fn resolve_relative_to(&self, path: PathBuf, relative_to: &Resource) -> Option<Resource> {
-        match relative_to {
-            Resource::Bundled(key) => {
-                // Bundled file can only refer to bundled files.
-                let relative_start_dir = parent_dir(PathBuf::from(key)).join(&path);
-                bundled(relative_start_dir)
-            }
-            Resource::FileSystem(relative_to_path) => {
-                if path.is_absolute() {
-                    return file_system(path);
-                }
-
-                let start_dir = parent_dir(relative_to_path.clone());
-                file_system(start_dir.join(&path)).or_else(|| {
-                    start_dir
-                        .strip_prefix(&self.root_dir)
-                        .ok()
-                        .and_then(|relative_start_dir| bundled(relative_start_dir.join(path)))
-                })
-            }
+    pub fn parent(&self) -> Option<&Path> {
+        match self {
+            Resource::FileSystem { file } => file.parent(),
+            _ => None,
         }
     }
 
-    pub fn exported(&self) -> bool {
-        BUNDLED
-            .file_names()
-            .any(|name| self.root_dir.join(name).exists())
+    pub fn absolutalize<P: AsRef<Path>>(&mut self, base: P) {
+        if let Resource::FileSystem { file: ref mut path } = self {
+            if path.is_relative() {
+                *path = base.as_ref().join(&path)
+            }
+        }
     }
 
-    pub fn export_ckb<'a>(&self, context: &TemplateContext<'a>) -> Result<()> {
-        self.export(CKB_CONFIG_FILE_NAME, context)
+    /// Gets resource content
+    pub fn get(&self) -> Result<Cow<'static, [u8]>> {
+        match self {
+            Resource::Bundled { bundled } => BUNDLED.get(bundled),
+            Resource::FileSystem { file } => {
+                let mut file = BufReader::new(fs::File::open(file)?);
+                let mut data = Vec::new();
+                file.read_to_end(&mut data)?;
+                Ok(Cow::Owned(data))
+            }
+        }
     }
 
-    pub fn export_miner<'a>(&self, context: &TemplateContext<'a>) -> Result<()> {
-        self.export(MINER_CONFIG_FILE_NAME, context)
+    /// Gets resource input stream
+    pub fn read(&self) -> Result<Box<dyn Read>> {
```

Comment:
unused function?

Reply:
It's a library, the library user can use when needed.